File: src/main/java/io/sc3/plethora/integration/vanilla/registry/VanillaMethodRegistration.java
package io.sc3.plethora.integration.vanilla.registry;

import net.minecraft.util.Identifier;
import io.sc3.plethora.api.method.IMethod;
import io.sc3.plethora.api.method.IMethodRegistry;
import io.sc3.plethora.api.module.IModuleContainer;
import io.sc3.plethora.integration.vanilla.method.EntityIntrospectionMethods;
import io.sc3.plethora.integration.vanilla.method.EntityKineticMethods;

public final class VanillaMethodRegistration {
    public static void registerMethods(IMethodRegistry r) {
        moduleMethod(r, "introspection:getInventory", EntityIntrospectionMethods.GET_INVENTORY);
        moduleMethod(r, "introspection:getEquipment", EntityIntrospectionMethods.GET_EQUIPMENT);
        moduleMethod(r, "introspection:getEnder", EntityIntrospectionMethods.GET_ENDER_CHEST);

        moduleMethod(r, "kinetic:look", EntityKineticMethods.LOOK);
        moduleMethod(r, "kinetic:use", EntityKineticMethods.USE);
        moduleMethod(r, "kinetic:swing", EntityKineticMethods.SWING);
    }

    private static <T> void method(IMethodRegistry r, String name, Class<T> target, IMethod<T> method) {
        r.registerMethod(Identifier.DEFAULT_NAMESPACE, name, target, method);
    }

    private static <T> void method(IMethodRegistry r, Class<T> target, IMethod<T> method) {
        r.registerMethod(Identifier.DEFAULT_NAMESPACE, method.getName(), target, method);
    }

    private static void moduleMethod(IMethodRegistry r, String name, IMethod<IModuleContainer> method) {
        method(r, name, IModuleContainer.class, method);
    }
}

File: src/main/java/io/sc3/plethora/integration/vanilla/registry/VanillaMethodRegistration.kt
package io.sc3.plethora.integration.vanilla.registry

import io.sc3.plethora.api.method.IMethod
import io.sc3.plethora.api.method.IMethodRegistry
import io.sc3.plethora.integration.vanilla.method.EntityIntrospectionMethods
import io.sc3.plethora.integration.vanilla.method.EntityKineticMethods
import io.sc3.plethora.integration.vanilla.method.RangedInventoryWrapperMethods
import net.minecraft.util.Identifier

object VanillaMethodRegistration {
  @JvmStatic
  fun registerMethods(registry: IMethodRegistry) {
    with (registry) {
      method("introspection:getInventory", EntityIntrospectionMethods.GET_INVENTORY)
      method("introspection:getEquipment", EntityIntrospectionMethods.GET_EQUIPMENT)
      method("introspection:getEnder", EntityIntrospectionMethods.GET_ENDER_CHEST)
      method("kinetic:look", EntityKineticMethods.LOOK)
      method("kinetic:use", EntityKineticMethods.USE)
      method("kinetic:swing", EntityKineticMethods.SWING)

      method("introspection:consume", RangedInventoryWrapperMethods.CONSUME)
      method("introspection:drop", RangedInventoryWrapperMethods.DROP)
    }
  }

  private inline fun <reified T> IMethodRegistry.method(name: String, method: IMethod<T>) {
    registerMethod(Identifier.DEFAULT_NAMESPACE, name, T::class.java, method)
  }
}
